### PR TITLE
CB-12755: Validating custom template to contains all only allowed services for upgrade.

### PR DESCRIPTION
**Pull request by tiborpopovics**

### CODE REVIEW

File: core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/image/BlueprintUpgradeOptionValidator.java [hunk 1]
```diff
@@ -42,7 +47,10 @@ private boolean isRuntimeUpgradeEnabled(Blueprint blueprint) {
     }
 
     private BlueprintUpgradeOption getBlueprintUpgradeOption(Blueprint blueprint) {
-        return Optional.ofNullable(blueprint.getBlueprintUpgradeOption())
-                .orElse(ENABLED);
+        return Optional.ofNullable(blueprint.getBlueprintUpgradeOption()).orElse(ENABLED);
+    }
+
+    private boolean isEnabledForCustomBlueprint(Blueprint blueprint) {
+        return !blueprint.getStatus().isDefault() && customTemplateUpgradeValidator.isValid(blueprint);
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/image/BlueprintUpgradeOptionValidator.java:54]: the first half of this seems to be always true

**Review by lacikaaa — CHANGES_REQUESTED** [review]
could you rephrase your commit message and maybe add some more details?
`contains all only allowed services for upgrade` is a bit ambiguous

**Review by lacikaaa — APPROVED** [review]

File: core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/image/CustomTemplateUpgradeValidator.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    private CmTemplateProcessorFactory cmTemplateProcessorFactory;
+
+    boolean isValid(Blueprint blueprint) {
+        if (isDatahubRuntimeUpgradeEnabledForCustomTemplate()) {
+            LOGGER.debug("Skipping custom template validation because the entitlement is turned on.");
+            return true;
+        } else {
+            Set<String> services = getServiceFromBlueprint(blueprint);
+            LOGGER.debug("Validating custom template. Permitted services for upgrade: {}, available services: {}", permittedServicesForUpgrade, services);
+            return permittedServicesForUpgrade.containsAll(services);
```
Comment (pdarvasi) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/image/CustomTemplateUpgradeValidator.java:41]: minor: we should log the result of the validation, too

File: core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/image/BlueprintUpgradeOptionValidator.java [hunk 3]
```diff
@@ -18,17 +20,20 @@
...
     private static final Logger LOGGER = LoggerFactory.getLogger(BlueprintUpgradeOptionValidator.class);
 
+    @Inject
+    private CustomTemplateUpgradeValidator customTemplateUpgradeValidator;
+
     boolean isValidBlueprint(Blueprint blueprint, boolean lockComponents) {
         LOGGER.debug("Validating blueprint upgrade option. Name: {}, type: {}, upgradeOption: {}, lockComponents: {}", blueprint.getName(),
                 blueprint.getStatus(), blueprint.getBlueprintUpgradeOption(), lockComponents);
-        return isCustomBlueprint(blueprint) || isEnabled(lockComponents, blueprint);
+        return isDefaultBlueprint(blueprint) ? isEnabledForDefaultBlueprint(lockComponents, blueprint) : isEnabledForCustomBlueprint(blueprint);
```
Comment (pdarvasi) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/image/BlueprintUpgradeOptionValidator.java:29]: this is a follow-up question, but don't we want to indicate why the blueprint is not valid for an upgrade to the end-user? cc: @keyki

File: core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/image/CustomTemplateUpgradeValidator.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+import com.sequenceiq.cloudbreak.cmtemplate.CmTemplateProcessorFactory;
+import com.sequenceiq.cloudbreak.domain.Blueprint;
+import com.sequenceiq.cloudbreak.template.model.ServiceComponent;
+
+@Component
+class CustomTemplateUpgradeValidator {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(CustomTemplateUpgradeValidator.class);
+
+    @Value("${cb.upgrade.permittedServicesForUpgrade:}")
```
Comment (keyki) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/image/CustomTemplateUpgradeValidator.java:25]: Can you please remove the `:`

File: core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/image/CustomTemplateUpgradeValidator.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            return permittedServicesForUpgrade.containsAll(services);
+        }
+    }
+
+    private boolean isDatahubRuntimeUpgradeEnabledForCustomTemplate() {
+        return entitlementService.datahubRuntimeUpgradeEnabledForCustomTemplate(Crn.safeFromString(ThreadBasedUserCrnProvider.getUserCrn())
+                .getAccountId());
+    }
+
+    private Set<String> getServiceFromBlueprint(Blueprint blueprint) {
```
Comment (keyki) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/image/CustomTemplateUpgradeValidator.java:50]: getService`s`

File: core/src/main/resources/application.yml [hunk 6]
```diff
@@ -712,6 +712,7 @@ cb:
         enabled: false
       sdx:
         enabled: false
+    permittedServicesForUpgrade: DAS,HDFS,HIVE,HIVE_ON_TEZ,HUE,KNOX,LIVY,LIVY_FOR_SPARK3,OOZIE,QUEUEMANAGER,SPARK3_ON_YARN,SPARK_ON_YARN,SQOOP_CLIENT,TEZ,YARN,ZEPPELIN,ZOOKEEPER
```
Comment (keyki) [inline core/src/main/resources/application.yml:715]: For better visibility, can we separate the services with new line?
